src/solver/batchnorm/forward_inference_ck.cpp

```diff
@@ -209,43 +218,63 @@ bool BnCKFwdInference::IsApplicable(
     return false;
 }
 
+template <typename InvokerFactoryMakerNHWC>
+ConvSolution MakeAnyInvokerFactory(const miopen::batchnorm::ProblemDescription& problem,
+                                   InvokerFactoryMakerNHWC&& invoker_factory_maker_nhwc)
+{
+#if MIOPEN_BACKEND_HIP && MIOPEN_USE_COMPOSABLEKERNEL
+    if(problem.IsLayoutNHWC())
+    {
+        switch(problem.GetXDesc().GetType())
+        {
+        case miopenFloat: return invoker_factory_maker_nhwc(F32{});
+        case miopenDouble: return invoker_factory_maker_nhwc(F64{});
+        case miopenHalf: return invoker_factory_maker_nhwc(F16{});
+        case miopenBFloat16: return invoker_factory_maker_nhwc(BF16{});
+        default:
+            MIOPEN_THROW(miopenStatusInternalError,
+                         "BnCKFwdInference operation does not support this data type");
+        }
+    }
+    // Todo: problem.IsLayoutDefault()
+    else
+    {
+        MIOPEN_THROW(miopenStatusInternalError,
+                     "BnCKFwdInference operation does not support this data layout");
+    }
+#else
+    return {};
+#endif
+}
+
 ConvSolution BnCKFwdInference::GetSolution(
     [[maybe_unused]] const ExecutionContext& context,
     [[maybe_unused]] const miopen::batchnorm::ProblemDescription& bn_problem) const
 {
 #if MIOPEN_BACKEND_HIP && MIOPEN_USE_COMPOSABLEKERNEL
-    ConvSolution result;
-    result.invoker_factory = [=](const std::vector<Kernel>& kernels) {
-        std::ignore = kernels;
-        return [=](const Handle& handle, const AnyInvokeParams& primitive_parameters) {
-            switch(bn_problem.GetXDesc().GetType())
+    return MakeAnyInvokerFactory(
+        bn_problem,
+        [&](auto data_type_val) {
+            using T = decltype(data_type_val);
+            if constexpr(std::is_same_v<T, F16>)
```

Comment:
This `if/else-if` logic can be simplified as follows: 
```
using AccTy = std::conditional_t<std::is_same_v<T, F64>, 
                               T, // pick this if true
                               F32>; // pick this if false
InvokerFactoryNHWC<T, T, AccT, T, T, AccT>(bn_problem); 
```

In fact InvokerFactory coudl also be changed to take just two template type parameters `T` and `AccT`.